Data-manager Migration from Play Framework to Dropwizard

igormoreno opened the pull request:
The migration consists mostly of:

- Migration from sbt to maven. This includes:
  - Support for running tests using scalatest
  - Support for incremental and continuos compilation and testing
  - Proper merging of akka configuration into fat jar
- Translation of controllers
- Translation of integration tests
- Simplifying dependences (for example in the case of logging)
- Changing data serialization from (json -> binary) to (json -> text)
  This will allow any application to restore the data if necessary

Still pending:

- [X] Migration of the configuration system
- [x] Proper dependency injection
- [ ] https://benchflow.slack.com/archives/C5D3KM6GK/p1495128171685736 (configuration.yml, GoogleDriveFactory)
- [ ] Document key steps of the migration
- [ ] Align Maven build options as in https://github.com/benchflow/benchflow/pull/495/

igormoreno committed:
data-manager: Migrate from Play framework to Dropwizard (#133, #83)

It consists mostly of:

- Migration from sbt to maven. This includes:
  - Support for running tests using scalatest
  - Support for incremental and continuos compilation and testing
  - Proper merging of akka configuration into fat jar
- Translation of controllers
- Translation of integration tests
- Simplifying dependences (for example in the case of logging)
- Changing data serialization from (json -> binary) to (json -> text)
  This will allow any application to restore the data if necessary

Still pending:

- Migration of the configuration system
- Proper dependency injection
- Document key steps of the migration

igormoreno committed:
Automatic scalastyle formatting

igormoreno committed:
Setting up scalastyle and scalariform and fixing style warnings

igormoreno committed:
Formatting

igormoreno committed:
Java style checker and formatter

Adding to maven a java style guideline checker:
https://maven.apache.org/plugins/maven-checkstyle-plugin/

It is set up to follow the google java style guidelines:
https://google.github.io/styleguide/javaguide.html

You can run it in isolation with
```
mvn checkstyle:check
```

In terms of the maven development lifecycle phases, these static
analysis tools are typically bound to the phase `validate` (which
happens before `compile`) or the phase `verify` (which happens after
`package`). I've looked online at several discussions on this topic. I
don't think neither of these choices are good. Using the `validate`
phase is bad because i think the 1st static check should be performed by
the compiler and if we use the `verify` phase we have to wait for the
entire packaging phase to happen and that takes quite some time. So i
decided to bind all the static analysis tools to the `test` phase.
```
mvn test
```

I'm working on adding other static checks that are also available in the
cadacy platform.

I also added to maven a java code formatter also following the google
java style guidelines. It will run automatically before compilation.
Formatting configuration files for IDEs are also
available: https://github.com/coveo/fmt-maven-plugin

I am using maven configuration inheritance to share the configuration of
these and other plug-ins among multiple projects

igormoreno committed:
Formatting. Following the Google Java style guidelines

igormoreno committed:
Trying to get codacy to read my configuration file

igormoreno committed:
Improving style

igormoreno committed:
Removing useless file

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

igormoreno committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

igormoreno committed:
Merge branch 'feature-dm-migration-dropwizard' of github.com:benchflow/benchflow into feature-dm-migration-dropwizard

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

igormoreno commented:
The most important pending items were added to the following issues:
- #278 
- #277 
- #292 
- #293

VincenzoFerme committed:
Merge branch 'devel' into feature-dm-migration-dropwizard

VincenzoFerme committed:
Removes dependency reduced pom

VincenzoFerme committed:
Removes Dockerfiles

VincenzoFerme committed:
Removes Makefile

VincenzoFerme committed:
Renames a file

VincenzoFerme committed:
Adds additional environment service

VincenzoFerme committed:
Adds build image script

VincenzoFerme committed:
Adds build and deploy pipeline for the data manager

VincenzoFerme committed:
Enables the build trigger for the data manager

VincenzoFerme commented:
@igormoreno I enabled CI build and Docker Image deployment for the service. The build fails, I guess the Test compliance described in https://github.com/benchflow/benchflow/issues/293 has to be applied before being able to merge the `data-manager`

jeppe-style commented:
Suggest to comply with API standard as defined in https://github.com/benchflow/benchflow/issues/293

Why do we have the package `service`? We don't have that in the other services.

Missing test that actually starts the application (see example in test-manager https://github.com/benchflow/benchflow/blob/devel/benchflow-test-manager/application/src/test/java/cloud/benchflow/testmanager/BenchFlowTestManagerApplicationIT.java)

Not all dependencies have the version set as a property.

Maybe also to add Swagger (see example in test-manager https://github.com/benchflow/benchflow/blob/devel/benchflow-test-manager/pom.xml#L130)

Probably also dropwizard-template-config is needed to read environment variables when running in docker (see example in test-manager https://github.com/benchflow/benchflow/blob/devel/benchflow-test-manager/pom.xml#L90)

jeppe-style commented on benchflow-data-manager/src/main/java/cloud/benchflow/datamanager/service/DataManagerConfiguration.java:
```diff
@@ -0,0 +1,33 @@
+package cloud.benchflow.datamanager.service;
```
should probably go into package `configurations` like the other services

igormoreno replied:
Sure. That's part of #293

igormoreno commented:
Where are these standards I have to comply with?

igormoreno committed:
Better application monitoring

igormoreno commented:
To do:

- [x] folder structure: #293 
- [ ] URL standards: #349 
- [x] configuration files: #277 
- [ ] integration tests: #348 
- [x] add swagger: #350
- [x] dependency version in properties: #351

igormoreno committed:
Merge branch 'develop' into feature-dm-migration-dropwizard

igormoreno committed:
Move version of dependencies to properties in POM (close #351)

igormoreno committed:
Automatic formatting

igormoreno committed:
Adapt folder structure to standard (#293)

igormoreno committed:
Remove unused guice annotations

igormoreno committed:
Instantiate Minio using dropwizard configuration (#277)

igormoreno committed:
Instantiate GoogleDrive using dropwizard config (#277)

igormoreno committed:
Towards standardized URLs (#349)

igormoreno committed:
Merge branch 'feature-dm-improvements' into feature-dm-migration-dropwizard

igormoreno committed:
Small bug fix

igormoreno committed:
Instantiate Cassandra using dropwizard config (closes #277)

igormoreno committed:
Merge branch 'develop' into feature-dm-migration-dropwizard

igormoreno committed:
Devops: make rule sets explicit to match what's in CI

igormoreno committed:
Fix issue: Fields should be declared at the top of the class

igormoreno committed:
Utility classes should be final and have private constructors

igormoreno committed:
Add swagger (closes #350)

igormoreno committed:
Various improvements to integration tests (closes #348)

- Starting entire application
- Injecting mocked dependences
- Starting docker container with Cassandra and accessing it from tests
- Running unit tests and integration tests separately

igormoreno commented:
I'm getting and error in wercker:
```
  Cause: java.lang.IllegalStateException: Could not find docker-compose, looked in: DockerCommandLocations{possiblePaths=[null, /usr/local/bin/docker-compose, /usr/bin/docker-compose]}
```

but i'm using "docker-compose-rule" which is also used in other projects. why doesn't it work?

VincenzoFerme commented:
Look in the following part of the `wrecker.yml` for the settings you need for Wrecker: https://github.com/benchflow/benchflow/blob/devel/wercker.yml#L114. In Wercker you can not interact with Docker Compose, so i came up with a flexible alternative. The code you took from @jeppe-style has also been modified by me to adapt it to Wercker.
